lpc55_sign/src/signed_image.rs

```diff
@@ -24,366 +22,123 @@ fn get_pad(val: usize) -> usize {
+    // goes into the image and _must_ match the hash-of-hashes programmed in
+    // the CMPA!
+    for root in root_certs {
+        image_bytes.extend_from_slice(&root_key_hash(root)?);
     }
-    // We may eventually have more hashes
-    out.write_all(&root0_sha)?;
-    out.write_all(&empty_hash)?;
-    out.write_all(&empty_hash)?;
-    out.write_all(&empty_hash)?;
-
-    // The sha256 of all the root keys gets put in in the CMPA area
-    let mut rkth_sha = sha2::Sha256::new();
-    rkth_sha.update(root0_sha);
-    rkth_sha.update(empty_hash);
-    rkth_sha.update(empty_hash);
-    rkth_sha.update(empty_hash);
-
-    let rkth = rkth_sha.finalize();
-
-    drop(out);
-
-    // the easiest way to get the bytes we need to sign is to read back
-    // what we just wrote
-    let sign_bytes = std::fs::read(outfile_path)?;
+    Ok(image_bytes)
+}
 
-    let mut img_hash = sha2::Sha256::new();
-    img_hash.update(&sign_bytes);
+/// Decode the private key, sign the stamped image with it,
+/// and append the signature to the image.
+pub fn sign_image(binary: &[u8], private_key: &str) -> Result<Vec<u8>> {
+    let mut image_hash = Sha256::new();
+    image_hash.update(binary);
 
-    let sig = priv_key.sign(
+    let private_key = RsaPrivateKey::from_pkcs1_pem(private_key)
+        .or_else(|_| RsaPrivateKey::from_pkcs8_pem(private_key))?;
+    let signature = private_key.sign(
         rsa::pkcs1v15::Pkcs1v15Sign::new::<rsa::sha2::Sha256>(),
-        img_hash.finalize().as_slice(),
+        image_hash.finalize().as_slice(),
     )?;
 
-    println!("Image signature {:x?}", sig);
-
-    let mut out = OpenOptions::new()
-        .write(true)
-        .append(true)
-        .open(outfile_path)?;
-
-    out.write_all(sig.as_ref())?;
-    drop(out);
-
-    Ok(rkth.as_slice().try_into().expect("something went wrong?"))
+    let mut signed = binary.to_owned();
+    signed.extend_from_slice(&signature);
+    Ok(signed)
 }
 
-pub fn create_cmpa(
-    with_dice: bool,
-    with_dice_inc_nxp_cfg: bool,
-    with_dice_cust_cfg: bool,
-    with_dice_inc_sec_epoch: bool,
-    rkth: &[u8; 32],
-    cmpa_path: &Path,
-) -> Result<()> {
-    let mut cmpa_out = OpenOptions::new()
-        .write(true)
-        .truncate(true)
-        .create(true)
-        .open(cmpa_path)?;
-
-    let mut secure_boot_cfg = SecureBootCfg::new();
-
-    secure_boot_cfg.set_dice(with_dice);
-    secure_boot_cfg.set_dice_inc_nxp_cfg(with_dice_inc_nxp_cfg);
-    secure_boot_cfg.set_dice_inc_cust_cfg(with_dice_cust_cfg);
-    secure_boot_cfg.set_dice_inc_sec_epoch(with_dice_inc_sec_epoch);
-    secure_boot_cfg.set_sec_boot(true);
-
-    let mut cmpa = CMPAPage::new();
-
-    cmpa.set_secure_boot_cfg(secure_boot_cfg)?;
-    cmpa.set_rotkh(rkth);
-
-    cmpa.set_debug_fields(DebugSettings::new())?;
-    cmpa.set_boot_cfg(DefaultIsp::Auto, BootSpeed::Fro96mhz)?;
-    let cmpa_bytes = cmpa.pack()?;
+pub fn root_key_hash(root: Vec<u8>) -> Result<[u8; 32]> {
```

Comment:
Nit: this could take `&[u8]` and avoid a copy

Reply:
Fixed in 3693dfa.